go/login/server/keymanager_test.go

```diff
@@ -0,0 +1,225 @@
+	k := map[string]*KeyManager{
+		"empty":   NewKeyManager(),
+		"DropAll": NewKeyManager(),
+		"Add":     NewKeyManager(),
+	}
+
+	k["DropAll"].DropAllReplace(
+		[]PrivateKey{
+			PrivateKey{
+				Value: glome.PrivateKey([32]byte{49, 244, 125, 133, 0, 40, 7,
+					192, 7, 90, 5, 208, 234, 104, 66, 68, 251, 237, 187, 132,
+					67, 236, 108, 164, 162, 199, 41, 89, 128, 95, 26, 190}),
+				Index: 122,
+			},
+			PrivateKey{
+				Value: glome.PrivateKey([32]byte{1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+					1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+					1, 1}),
+				Index: 123,
+			},
+		})
+
+	k["Add"].Add(glome.PrivateKey([32]byte{1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+		1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+		1, 1}), 123)
+
+	return k
+}
+
+func Find(list []PublicKey, pub PublicKey) bool {
+	for _, b := range list {
+		if b == pub {
+			return true
+		}
+	}
+	return false
+}
+
+func TestKeyAdd(t *testing.T) {
+	for name, manager := range tests() {
+		name := name
+		manager := manager
+
+		t.Run(name, func(t *testing.T) {
+			for _, k := range []struct {
+				priv  glome.PrivateKey
+				index uint8
+			}{
+				{
+					priv:  glome.PrivateKey([32]byte{}),
+					index: 0,
+				}, {
+					priv: glome.PrivateKey([32]byte{1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+						1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+						1, 1}),
+					index: 12,
+				}, {
+					priv: glome.PrivateKey([32]byte{49, 244, 125, 133, 0, 40, 7,
+						192, 7, 90, 5, 208, 234, 104, 66, 68, 251, 237, 187, 132,
+						67, 236, 108, 164, 162, 199, 41, 89, 128, 95, 26, 190}),
+					index: 23,
+				},
+			} {
+				if err := manager.Add(k.priv, k.index); err != nil {
+					t.Fatalf("test %v, unexpected error: %v ", name, err.Error())
+				}
+
+				if manager.indexToPriv[k.index] != k.priv {
+					t.Errorf("test %v: private key %v was not added in index %v",
+						name, k.priv, k.index)
+				}
+
+				pub, err := k.priv.Public()
+				if err != nil {
+					t.Fatalf("test %v, unexpected error: %v ", name, err.Error())
+				}
+
+				if !Find(manager.publicKeys, PublicKey{Value: *pub, Index: k.index}) {
+					t.Errorf("test %v: public key %v was not added in index %v",
+						name, pub, k.index)
+				}
+			}
+		})
+	}
+}
+
+func TestKeyAddExceptions(t *testing.T) {
+	for name, manager := range tests() {
+		name := name
+		manager := manager
+
+		type input struct {
+			priv  glome.PrivateKey
+			index uint8
+		}
+
+		t.Run(name, func(t *testing.T) {
+			for _, k := range []struct {
+				in   input
+				want error
+			}{
+				{in: input{priv: glome.PrivateKey([32]byte{}), index: 0}, want: nil},
+				{in: input{priv: glome.PrivateKey([32]byte{}), index: 0}, want: ErrOverloadedIndex{index: 0}},
+				{in: input{priv: glome.PrivateKey([32]byte{}), index: 129}, want: ErrInvalidIndex{index: 129}},
+			} {
+				if err := manager.Add(k.in.priv, k.in.index); err != k.want {
+					t.Errorf("%v failed; got %#v, want %#v", name, err, k.want)
```

Comment:
The test itself already contains `name`, you may want to log the index of the key to add here.

Reply:
Fixed